### PR TITLE
GTEST/UCP: Decrease memtype length size under valgrind

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Decrease memtype length size when running under valgrind.

## Why ?

before:
```
[ RUN      ] udx/test_ucp_tag_mem_type.basic/0 <ud_x,cuda_copy,rocm_copy/host:host>
[       OK ] udx/test_ucp_tag_mem_type.basic/0 (6701 ms)
```
after:
```
[ RUN      ] udx/test_ucp_tag_mem_type.basic/0 <ud_x,cuda_copy,rocm_copy/host:host>
[       OK ] udx/test_ucp_tag_mem_type.basic/0 (1366 ms)
```
memtype tests are one of the longest tests:
```
2021-09-13T20:03:09.6520202Z TOP-20 longest tests:
...
2021-09-13T20:03:09.6707240Z 15. udx/test_ucp_tag_mem_type.basic/40                                                                                        - 15243 ms
2021-09-13T20:03:09.6716464Z 16. ud/test_ucp_tag_mem_type.basic/68                                                                                         - 14153 ms
2021-09-13T20:03:09.6725536Z 17. udx/test_ucp_tag_mem_type.basic/68                                                                                        - 13992 ms
...
```

## How ?

1. Introduce auxiliary methods: `max_test_length()` and `test_length()`.
2. Add dividing `max_length` by `ucs::test_time_multiplier()` in `max_test_length()`.
3. Use the functions to reduce duplication and use smaller size when running under valgrind.

### CODE REVIEW

File: test/gtest/ucp/test_ucp_tag_mem_type.cc [hunk 1]
```diff
@@ -94,6 +94,17 @@ class test_ucp_tag_mem_type: public test_ucp_tag {
         mem_buffer::pattern_check(rbuf, length, 2, r_mem_type);
     }
 
+    size_t max_test_length(unsigned exp) const
+    {
+        size_t max_length = static_cast<size_t>(pow(10.0, exp));
+        return ucs_max(10, max_length % ucs::test_time_multiplier());
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_mem_type.cc:100]: i think better reduce num of iterations. to not lose coverage

Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc:100]: but here we don't have iterations, we go through `i=[1..7]` and calculate message size as `pow(10, i)`

Comment (yosefe) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc:100]: iterate 1..3 and calculate size pow(10,i*2)?

Comment (brminich) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc:100]: maybe do not change it at all, as the benefit does not look significant enough

Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc:100]: we do all available pairs of host,cuda,cuda-managed,rocm,rocm-managed, i.e. 5^5=125

Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc]: > iterate 1..3 and calculate size pow(10,i*2)?

done

File: test/gtest/ucp/test_ucp_tag_mem_type.cc [hunk 2]
```diff
@@ -176,9 +175,11 @@ size_t test_ucp_tag_mem_type::do_xfer(const void *sendbuf, void *recvbuf,
 
 UCS_TEST_P(test_ucp_tag_mem_type, basic)
 {
-    ucp_datatype_t type = ucp_dt_make_contig(1);
+    const size_t max_iter       = RUNNING_ON_VALGRIND ? 3 : 7; 
+    const size_t pow_multiplier = RUNNING_ON_VALGRIND ? 2 : 1; 
+    ucp_datatype_t type   = ucp_dt_make_contig(1);
 
-    for (unsigned i = 1; i <= 7; ++i) {
+    for (unsigned i = 1; i <= max_iter; i *= pow_multiplier) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_mem_type.cc:182]: should be `i+=`, not `i*=`

File: test/gtest/ucp/test_ucp_tag_mem_type.cc [hunk 3]
```diff
@@ -176,9 +175,11 @@ size_t test_ucp_tag_mem_type::do_xfer(const void *sendbuf, void *recvbuf,
 
 UCS_TEST_P(test_ucp_tag_mem_type, basic)
 {
-    ucp_datatype_t type = ucp_dt_make_contig(1);
+    const size_t max_iter       = RUNNING_ON_VALGRIND ? 3 : 7; 
+    const size_t pow_multiplier = RUNNING_ON_VALGRIND ? 2 : 1; 
+    ucp_datatype_t type   = ucp_dt_make_contig(1);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_mem_type.cc:180]: align =
make it const?

Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc]: done

File: test/gtest/ucp/test_ucp_tag_mem_type.cc (hunk 2 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc]: good catch, done

**Comment by dmitrygx** [comment]
@yosefe is it ok now?

File: test/gtest/ucp/test_ucp_tag_mem_type.cc [hunk 4]
```diff
@@ -94,6 +94,16 @@ class test_ucp_tag_mem_type: public test_ucp_tag {
...
     }
 
+    size_t max_test_length(unsigned exp) const
+    {
+        return static_cast<size_t>(pow(10.0, exp));
+    }
+
+    size_t test_length(size_t max_length) const
+    {
+        return ucs::rand() % max_length + 1;
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_mem_type.cc:104]: add ( )

File: test/gtest/ucp/test_ucp_tag_mem_type.cc [hunk 5]
```diff
@@ -165,12 +175,12 @@ size_t test_ucp_tag_mem_type::do_xfer(const void *sendbuf, void *recvbuf,
...
-    size_t max_length;
+    const size_t max_iter1       = RUNNING_ON_VALGRIND ? 3 : 7; 
+    const size_t iter_multiplier = RUNNING_ON_VALGRIND ? 2 : 1; 
+    const ucp_datatype_t type    = ucp_dt_make_contig(1);
 
-    for (unsigned i = 1; i <= 7; ++i) {
-        max_length = (size_t)pow(10.0, i);
-        size_t length = ucs::rand() % max_length + 1;
+    for (unsigned i = 1; i <= max_iter1; ++i) {
+        size_t length = test_length(max_test_length(i * iter_multiplier));
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_mem_type.cc:183]: maybe call max_test_length from test_length?

File: test/gtest/ucp/test_ucp_tag_mem_type.cc [hunk 6]
```diff
@@ -180,11 +190,12 @@ UCS_TEST_P(test_ucp_tag_mem_type, basic)
     }
 
     /*  test with re-using the buffers */
-    max_length = (size_t)pow(10.0, 7);
+    size_t max_length = test_length(max_test_length(7));
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_tag_mem_type.cc:193]: why need test_length wrapper here?

Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc]: fixed

File: test/gtest/ucp/test_ucp_tag_mem_type.cc (hunk 5 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc]: done

File: test/gtest/ucp/test_ucp_tag_mem_type.cc (hunk 4 above)
Comment (dmitrygx) [reply test/gtest/ucp/test_ucp_tag_mem_type.cc]: done

**Review by yosefe — APPROVED** [review]

**Pull request merged**